File: python/operating-on-numbers-using-container/callerService/caller.py
from flask import Flask,request,jsonify
from flask_cors import CORS
import requests
from opentracing.ext import tags
from opentracing.propagation import Format
import logging
from jaeger_client import Config

app = Flask(__name__)
CORS(app)

def init_tracer(service):
    logging.getLogger('').handlers = []
    logging.basicConfig(format='%(message)s', level=logging.DEBUG)

    config = Config(
        config={
            'sampler': {
                'type': 'const',
                'param': 1,
            },
            'local_agent': {
                'reporting_host': 'jaeger',
                'reporting_port': 6831
            },
            'logging': True,
        },
        service_name=service,
    )

    # this call also sets opentracing.tracer
    return config.initialize_tracer()

tracer = init_tracer('CallerService') 

@app.route("/",methods=['POST'])
def caller():
    span = tracer.start_span("Caller")
    payload = request.get_json()
    sqrurl = 'http://sqrapp:5020/'
    # trace_id
    # span.set_tag('Trace_id', span.context.trace_id)
    print('-----------span context---------')
    print(span.context)
    span.set_tag(tags.HTTP_METHOD, 'POST')
    span.set_tag(tags.SPAN_KIND, tags.SPAN_KIND_RPC_CLIENT)
    span.set_tag(tags.HTTP_URL, sqrurl)
    headers = {'Content-Type': 'application/json'}
    tracer.inject(span, Format.HTTP_HEADERS, headers)
    # print("headers------------",headers)
    sqrres = requests.post(sqrurl,json=payload,headers=headers)
    sqrdata = sqrres.json()
    sqrnum1 = int(sqrdata['num1'])
    sqrnum2 = int(sqrdata['num2'])

    cubeurl = 'http://cubeapp:5030/'
    span.set_tag(tags.HTTP_URL, cubeurl)
    # headers = {'Content-Type': 'application/json'}
    # tracer.inject(span, Format.HTTP_HEADERS, headers)
    cuberes = requests.post(cubeurl,json=payload,headers=headers)
    cubedata = cuberes.json()
    cubenum1 = int(cubedata['num1'])
    cubenum2 = int(cubedata['num2'])

    
    finalpayload = {
        "num1" : sqrnum1,
        "num2" : sqrnum2,
        "num3" : cubenum1,
        "num4" : cubenum2
    }

    addurl = 'http://addapp:5040/'
    span.set_tag(tags.HTTP_URL, addurl)
    # headers = {'Content-Type': 'application/json'}
    # tracer.inject(span, Format.HTTP_HEADERS, headers)
    addres = requests.post(addurl,json=finalpayload,headers=headers)
    add_data = addres.json()
    addresult = add_data['result']

    span.finish()
    return jsonify(sqrnum1=sqrnum1,
    sqrnum2=sqrnum2, cubenum1=cubenum1, 
    cubenum2=cubenum2, theresult=addresult)

if __name__ == '__main__':
    app.run(host="0.0.0.0",port=5010,debug=True)

File: nodejs/operatingOnNumbers/caller/caller.js
const express = require('express');
const bodyParser = require('body-parser');
const axios = require('axios');
const initJaegerTracer = require("jaeger-client").initTracer;
const { Tags, FORMAT_HTTP_HEADERS } = require('opentracing');

const app = express();
const port = 9000;
app.use(bodyParser.json()); // support json encoded bodies
app.use(bodyParser.urlencoded({ extended: true })); // support encoded bodies

function initTracer(serviceName) {
  const config = {
    serviceName: serviceName,
    sampler: {
      type: "const",
      param: 1,
    },
    reporter: {
      agentHost: 'jaeger',
      agentPort: 6832,
      logSpans: true,
    },
  };
  const options = {
    logger: {
      info(msg) {
        console.log("INFO ", msg);
      },
      error(msg) {
        console.log("ERROR", msg);
      },
    },
  };
  return initJaegerTracer(config, options);
}

const tracer = initTracer("caller");

app.post("/", (req,res)=>{
  const span = tracer.startSpan("call");
  var num1 = req.body.num1;
  var num2 = req.body.num2;
  const sqrurl = 'http://sqrnode:9001';
  const cubeurl = 'http://cubenode:9002';
  const addurl = 'http://addnode:9003';
  const headers = {'Content-Type': 'application/json'};
  span.setTag(Tags.HTTP_METHOD, 'POST');
  span.setTag(Tags.SPAN_KIND, Tags.SPAN_KIND_RPC_CLIENT);
  span.setTag(Tags.HTTP_URL, sqrurl);
  tracer.inject(span, FORMAT_HTTP_HEADERS, headers);
  axios.post(sqrurl, {
      num1 : num1,
      num2 : num2
    },
    {
      headers : headers
    })
    .then(function (response) {
      sqrnum1 = response.data.sqrnum1;
      sqrnum2 = response.data.sqrnum2;
      console.log(response.data);

      span.setTag(Tags.HTTP_URL, cubeurl);
      axios.post(cubeurl, {
        num1 : num1,
        num2 : num2,
      },
      {
        headers : headers
      })
      .then (function(response1) {
        cubenum1 = response1.data.cubenum1;
        cubenum2 = response1.data.cubenum2;
        console.log(response1.data);
        
        span.setTag(Tags.HTTP_URL, addurl);
        axios.post(addurl, {
          num1 : sqrnum1,
          num2 : sqrnum2,
          num3 : cubenum1,
          num4 : cubenum2
        },
        {
          headers : headers
        })
        .then (function(response2){
          result = response2.data.result;
          console.log(response2.data);

          span.finish();
          res.send(JSON.stringify({sqrnum1:sqrnum1,sqrnum2:sqrnum2,cubenum1:cubenum1,
          cubenum2:cubenum2,result:result}));
        })
        .catch(function(error2){
          console.log(error2);
        })
      })
      .catch(function (error1){
        console.log(error1);
      })
    })
    .catch(function (error) {
      console.log(error);
    });
})


app.listen(port , ()=>{
    console.log("Listening to 9000");
})

// app.post("/sqr", (req,res)=>{
//   var num1 = req.body.num1;
//   var num2 = req.body.num2;

//   axios.post('http://localhost:9001', {
//       num1 : num1,
//       num2 : num2
//     })
//     .then(function (response) {
//       sqrnum1 = response.data.sqrnum1;
//       sqrnum2 = response.data.sqrnum2;
//       console.log(response.data);
//       res.send(JSON.stringify({sqrnum1: sqrnum1, sqrnum2: sqrnum2}));
//     })
//     .catch(function (error) {
//       console.log(error);
//     });
// })

// app.post("/cube", (req,res)=>{
//     var num1 = req.body.num1;
//     var num2 = req.body.num2;

//     axios.post('http://localhost:9002', {
//         num1 : num1,
//         num2 : num2
//       })
//       .then(function (response) {
//         cubenum1 = response.data.cubenum1;
//         cubenum2 = response.data.cubenum2;
//         console.log(response.data);
//         res.send(JSON.stringify({cubenum1: cubenum1, cubenum2: cubenum2}));
//       })
//       .catch(function (error) {
//         console.log(error);
//       });
// })

// app.post("/add", (req,res)=>{
//     var num1 = req.body.num1;
//     var num2 = req.body.num2;
//     var num3 = req.body.num3;
//     var num4 = req.body.num4;

//     axios.post('http://localhost:9003', {
//         num1 : num1,
//         num2 : num2,
//         num3 : num3,
//         num4 : num4
//       })
//       .then(function (response) {
//         var addres = response.data.result;
//         console.log(response.data.result);
//         res.send(JSON.stringify({result : addres}));
//       })
//       .catch(function (error) {
//         console.log(error);
//       });
// })




